### PR TITLE
Origin/filters

### PR DESCRIPTION
**Pull request by DeeaMatei**
#### Description
This feature adds Filters to the Freaks List. The filters are Skills and Projects. The filtration is not yet implemented (waiting for data from backend).

#### [Task](task_url)
https://trello.com/c/lZv0aAXB/41-android-freaks-screen-filter-freaks-by-skills-and-projects

### CODE REVIEW

**Comment by DeeaMatei** [comment]
[Android: Freaks Screen - Filter Freaks by skills and projects](https://trello.com/c/lZv0aAXB/41-android-freaks-screen-filter-freaks-by-skills-and-projects)

File: features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FilterAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        }
+    }
+
+    @SuppressLint("NotifyDataSetChanged")
+    fun resetCheckboxes(){
+        reset = true
+        notifyDataSetChanged()
+    }
+
+
```
Comment (cipry13) [inline features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FilterAdapter.kt]: Trailing whitespace

File: features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt [hunk 2]
```diff
@@ -83,4 +98,38 @@ class FreaksFragment : Fragment() {
...
     }
+
+    private fun loadFilters(activeFilter: String) =
+        if (activeFilter == "Skills") {
+            listOf<String>("Android", "Kotlin", "Other Skill", "iOS")
+        } else {
+            listOf("Freaks Catalog", "Proj2", "Tutorial", "Altkeva")
+        }
+
+    private fun showFilterModal(activeFilter: String, otherView: ViewGroup?){
```
Comment (cipry13) [inline features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt]: typo in Modal

File: features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FilterAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    override fun onBindViewHolder(holder: FilterViewHolder, position: Int) {
+        val filter = filters[position]
+        holder.checkBox.text = filter
+        if (reset) {
+            holder.checkBox.isChecked = false
+        }
+    }
+
+    @SuppressLint("NotifyDataSetChanged")
```
Comment (cipry13) [inline features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FilterAdapter.kt]: Is this the only way do to it (SurpressLint) ?

File: features/freaks/src/main/res/layout/bottom_sheet_dialog.xml [hunk 4]
```diff
@@ -0,0 +1,63 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical"
+    app:layout_behavior="com.google.android.material.bottomsheet.BottomSheetBehavior">
+
+    <LinearLayout
+        android:id="@+id/checkboxes"
```
Comment (cipry13) [inline features/freaks/src/main/res/layout/bottom_sheet_dialog.xml]: Could have picked a better name (maybe checkboxes_linear_layout?)

File: features/freaks/src/main/res/layout/filter_item.xml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+android:layout_width="match_parent"
+android:layout_height="wrap_content"
+android:orientation="vertical">
+
+    <View
+        android:layout_width="380dp"
+        android:layout_height="2dp"
+        android:layout_marginStart="15dp"
+        android:layout_marginTop="10dp"
+        android:background="@color/gray" />
```
Comment (cipry13) [inline features/freaks/src/main/res/layout/filter_item.xml]: Is this better than adding the divider with addItemDecoration?

File: features/freaks/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -1,3 +1,11 @@
 <resources>
     <string name="freaks_title">Freaks screen</string>
+    <string name="skills_btn">Skills</string>
+    <string name="projects_btn">Projects</string>
+    <string name="resetBtn">Reset</string>
+    <string name="androidCheckBox">Android</string>
+    <string name="iosCheckBox">iOS</string>
+    <string name="cevaCheckBox">Ceva</string>
+    <string name="altcevaCheckBox">Altceva</string>
```
Comment (cipry13) [inline features/freaks/src/main/res/values/strings.xml]: Stuff used for testing should be removed

File: features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt [hunk 7]
```diff
@@ -83,4 +98,38 @@ class FreaksFragment : Fragment() {
...
+        dialog.setContentView(view)
+        dialog.show()
+        btReset?.setOnClickListener {
+            myAdapter.resetCheckboxes()
+        }
+
+        btApply?.setOnClickListener {
+
+        }
+    }
```
Comment (cipry13) [inline features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt]: Seems a bit messy, maybe move BottomSheetDialog related code in a new Class?

**Review by cipry13 — COMMENTED** [review]
❌

**Review by cipry13 — APPROVED** [review]

**Review by srobert99 — APPROVED** [review]

File: build.gradle.kts [hunk 8]
```diff
@@ -12,6 +12,7 @@ allprojects {
 }
 
 buildscript {
+    val kotlin_version by extra("1.5.21")
```
Comment (AlexandraDamaschin) [inline build.gradle.kts]: Why did you add this here?

File: features/freaks/src/main/AndroidManifest.xml [hunk 9]
```diff
@@ -8,6 +9,12 @@
         <dist:delivery>
             <dist:install-time />
         </dist:delivery>
+
         <dist:fusing dist:include="true" />
     </dist:module>
-</manifest>
+
+    <application>
+        <activity android:name=".TestActivity"></activity>
```
Comment (AlexandraDamaschin) [inline features/freaks/src/main/AndroidManifest.xml]: Why do we need this Test Activity?

File: features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FilterAdapter.kt [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+import android.view.ViewGroup
+import android.widget.CheckBox
+import androidx.recyclerview.widget.RecyclerView
+
+class FilterAdapter(private val filters: List<String>) :
+    RecyclerView.Adapter<FilterAdapter.FilterViewHolder>() {
+    var reset: Boolean = false
+
+    class FilterViewHolder(view: View) : RecyclerView.ViewHolder(view) {
+        val checkBox: CheckBox = view.findViewById(R.id.skill)
```
Comment (AlexandraDamaschin) [inline features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FilterAdapter.kt]: checkBox it's a very general name, isn't this more as a skills_checkbox?

File: features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FilterAdapter.kt [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FilterViewHolder {
+        val adapterLayout =
+            LayoutInflater.from(parent.context).inflate(R.layout.filter_item, parent, false)
+        return FilterViewHolder(adapterLayout)
+    }
+
+    override fun onBindViewHolder(holder: FilterViewHolder, position: Int) {
+        val filter = filters[position]
+        holder.checkBox.text = filter
+        if (reset) {
```
Comment (AlexandraDamaschin) [inline features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FilterAdapter.kt]: What does reset means?

File: features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt [hunk 12]
```diff
@@ -49,6 +54,16 @@ class FreaksFragment : Fragment() {
         recyclerView.layoutManager = layoutManager
         recyclerView.adapter = ItemAdapter(freaksList)
 
+        val btShowSkills: Button = viewBinding.skillsBtn
```
Comment (AlexandraDamaschin) [inline features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt]: What means btn?

File: features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt [hunk 13]
```diff
@@ -49,6 +54,16 @@ class FreaksFragment : Fragment() {
...
+        val btShowProjects: Button = viewBinding.projectsBtn
+
+        btShowSkills.setOnClickListener {
+            showFilterModal("Skills", null)
```
Comment (AlexandraDamaschin) [inline features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt]: Hardcoded values in the code aren't a good idea, what do you think?

File: features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt [hunk 14]
```diff
@@ -83,4 +98,43 @@ class FreaksFragment : Fragment() {
             repeat(FREAKS_COUNT) { this.add(freak) }
         }
     }
+
+    private fun loadFilters(activeFilter: String) =
+        if (activeFilter == "Skills") {
```
Comment (AlexandraDamaschin) [inline features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt]: What happens if by mistake instead of "Skills" I pus "Skill"? Because I am just a simple person who can always make keyboard mistakes

File: features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt (hunk 2 above)
Comment (AlexandraDamaschin) [reply features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt]: There is a warning over otherView. We should always make sure to resolve all this warnings android studio gives us

File: features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt [hunk 15]
```diff
@@ -83,4 +98,43 @@ class FreaksFragment : Fragment() {
...
+        val filterTitle: TextView? = view.findViewById(R.id.filter_title)
+        filterTitle?.text = activeFilter
+        val btReset: TextView? = view.findViewById(R.id.reset)
+        val btApply: Button? = view.findViewById(R.id.apply_btn)
+        val recyclerFiltersView = view?.findViewById<RecyclerView>(R.id.recycler_filters_view)
+        val filtersList = loadFilters(activeFilter)
+        recyclerFiltersView?.layoutManager = LinearLayoutManager(view.context)
+
+        val dividerItemDecoration = DividerItemDecoration(
+            recyclerFiltersView?.getContext(),
```
Comment (AlexandraDamaschin) [inline features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt]: Android studio says: Use of getter method instead of property access syntax 
I really believe he is right, what do you think?

File: features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt [hunk 16]
```diff
@@ -83,4 +98,43 @@ class FreaksFragment : Fragment() {
...
+            (recyclerFiltersView?.layoutManager as LinearLayoutManager).getOrientation()
```
Comment (AlexandraDamaschin) [inline features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt]: Android studio says: Use of getter method instead of property access syntax
I really believe he is right, what do you think?

File: features/freaks/src/main/res/layout/bottom_sheet_dialog.xml [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+                android:paddingEnd="15dp"
+                android:text="@string/resetBtn"
+                app:layout_constraintEnd_toEndOf="parent"
+                app:layout_constraintTop_toTopOf="parent" />
+        </androidx.constraintlayout.widget.ConstraintLayout>
+    </LinearLayout>
+
+    <Button
+        android:id="@+id/apply_btn"
+        android:layout_width="150dp"
```
Comment (AlexandraDamaschin) [inline features/freaks/src/main/res/layout/bottom_sheet_dialog.xml]: Hardcoded values

File: features/freaks/src/main/res/layout/bottom_sheet_dialog.xml [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+        android:layout_height="50dp"
```
Comment (AlexandraDamaschin) [inline features/freaks/src/main/res/layout/bottom_sheet_dialog.xml]: Harcoded values all over the code

File: features/freaks/src/main/res/layout/filter_item.xml [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+android:layout_width="match_parent"
+android:layout_height="wrap_content"
+android:orientation="vertical">
+
+    <CheckBox
+        android:id="@+id/skill"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:checked="false"
+        android:buttonTint="#105FBF"
```
Comment (AlexandraDamaschin) [inline features/freaks/src/main/res/layout/filter_item.xml]: Hardcoded colors

**Review by AlexandraDamaschin — CHANGES_REQUESTED** [review]
There are a few things which need improvement. Take your time to fix and reflect on them.

File: features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FilterAdapter.kt (hunk 11 above)
Comment (DeeaMatei) [reply features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FilterAdapter.kt]: Reset is a Boolean that shows if the Reset Filters buttons is pressed. I changed it's name to filterReset to be more obvious.

File: features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt (hunk 14 above)
Comment (DeeaMatei) [reply features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt]: No one is actually typing what kind of filter they want to see. We pass a string with the filter name so we can also set the Filter Title when the bottom sheet dialog shows up. I will think of better ways to do this anyway.

File: features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt (hunk 15 above)
Comment (DeeaMatei) [reply features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt]: He is. I changed it to property access.

File: build.gradle.kts (hunk 8 above)
Comment (DeeaMatei) [reply build.gradle.kts]: I didn't. But it seems like we didn't use it anyway so I removed it.

File: features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt (hunk 12 above)
Comment (DeeaMatei) [reply features/freaks/src/main/kotlin/com/agilefreaks.freaks_catalog.features.freaks/FreaksFragment.kt]: It was supposed to show that it's a button. I changed it's name to showSkillsButton for it to be more obvious.

**Review by AlexandraDamaschin — APPROVED** [review]
Does this improve software quality? Yes
